### PR TITLE
Home page fans

### PR DESCRIPTION
**Pull request by Bashkir15**
The fans home page is styled closely to Basil's mockup and is completely responsive. I wasn't given the fonts, so I just took a wild guess and grabbed something similar from Google fonts, but that should be easy enough to swap out. The first signup modal trigger has a neat little hover effect. 

I also added in a gulp task simply to compile sass because writing plain css makes me want to pull my hair out. Feel free to remove it or switch it to whatever task runner you like. I just chunked everything into a main.css file, but the directory structure of the actual sass files permits for making the files more component friendly, like a directory for fan pages and such that way if we did get going in React it would be very easy to just compile all of the fan sass into css and then load in the css module, or css file with rollup, into a wrapper that will handle conditional rendering for the fan pages and things of that mindset, 

I also added in my personal reset/normalize css that I use for pretty much anything. Change it, take it out, hate it, state concerns blah blah blah. 

Included a nav, but that needs to be extracted out into hb templates to save our fingers. 

** The page is not 100% accurate with the mockups due to me missing one icon and the incorrect header background image, but it's close enough.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -55,7 +55,15 @@
   },
   "devDependencies": {
     "ava": "^0.18.1",
+    "browser-sync": "^2.18.8",
     "eslint-config-tidy": "^0.4.0",
+    "gulp": "^3.9.1",
```
Comment (sholladay) [inline package.json]: If we are going to use `gulp`, it needs to be Gulp 4. Too much cruft in the old ecosystem.

Replace `^3.9.1` with `gulpjs/gulp#4.0`. Note that this will probably break your stuff. You'll need to update your code and the other dependencies. The only one I know off the top of my head is that `gulp-plumber` is unnecessary now, but there are probably others that need to be replaced.

File: lib/view/home-pages/fan-home.html [hunk 2]
```diff
@@ -0,0 +1,197 @@
+<!DOCTYPE html>
```
Comment (sholladay) [inline lib/view/home-pages/fan-home.html]: The Handlebars layouts are responsible for the `<html>` boilerplate. See `lib/view/layout/default-layout`. When your view gets rendered, it will be inside of that template.

File: lib/static/css/common.css [hunk 3]
```diff
@@ -2,6 +2,13 @@
     margin  : 0;
     padding : 0;
 }
+
+*,
+*::before,
+*::after {
+    box-sizing: border-box;
```
Comment (sholladay) [inline lib/static/css/common.css]: This tells the rendering engine to behave like Internet Explorer...

https://developer.mozilla.org/en-US/docs/Web/CSS/box-sizing#Values

File: gulpfile.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+const gulp = require('gulp');
+const plumber = require('gulp-plumber');
+const notify = require('gulp-notify');
+const gutil = require('gulp-util');
+
+const sass = require('gulp-sass');
+const pure = require('gulp-purifycss');
+const autoprefixer = require('gulp-autoprefixer'); // incase we want to be nice
```
Comment (sholladay) [inline gulpfile.js]: I'm a bit wary of code transformation. What feature do you want to use that is currently prefixed in the latest browsers? A lot of things don't require vendor prefixes anymore, though not everything. If there is nothing in particular that you want to use that requires a prefix, I'd rather remove this.

File: lib/view/home-pages/fan-home.html [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		<meta name='viewport' content="width=device-width, initial-scale=1">
+
+		<title>Noirdoor</title>
+
+		<link href="https://fonts.googleapis.com/css?family=Open+Sans:300,400,600|Raleway:400,500,600" rel="stylesheet">
+		<link rel='stylesheet' href='../../static/css/compiled/main.css'>
+	</head>
+
+	<body>
+		<nav class='nav'>
```
Comment (sholladay) [inline lib/view/home-pages/fan-home.html]: The nav bar already exists here: https://github.com/noirdoor/noirdoor/blob/fab5096b197288c2d73101eec4e49446ae0d6307/lib/view/partial/header.html

File: package.json (hunk 1 above)
Comment (Bashkir15) [inline package.json]: Oops. I thought I cleaned up all of those. I'll fix that.

Okay, cool. I much prefer gulp 4, but I didn't want to pull something in that others hadn't used before and throw them for a loop. Will update to gulp 4

File: lib/static/css/common.css (hunk 3 above)
Comment (Bashkir15) [inline lib/static/css/common.css]: Well, behave like quirks mode in IE. box-sizing: border-box is pretty much a defacto standard these days. It makes spacing much more reasonable to deal and reason with. You specify a width and then that width doesn't change. If you add padding to the element then the content on the inside of the box is reduced by padding, but the width of the box doesn't change. Where as if you set a divs width to be 250px and then added 20px of padding to both sizes it would render as 290px in width.

Using this makes things like interpolation using calc() to use a completely cohesive design between individual pixels in resolutions instead of arbitrarily defined media queries.

We can talk about this more, but this should stay.

File: lib/view/home-pages/fan-home.html (hunk 5 above)
Comment (Bashkir15) [inline lib/view/home-pages/fan-home.html]: Hmm, there appears to have been some miscommunication. I started building based on the mockup that I received from Basil here. 

https://xd.adobe.com/view/cd941c2a-1b7f-4143-9f2a-8efc445f24fc/screen/ab6e5ef8-133d-4e56-a0b0-403a8db54fc0/HOME-PAGE-Fans

I didn't want to mess with things that someone else had already done so I figured I'd just include it and get feedback on what it will look like vs. the design I received

File: lib/view/home-pages/fan-home.html (hunk 2 above)
Comment (Bashkir15) [inline lib/view/home-pages/fan-home.html]: I figured as much, but was wary about just throwing my custom stylesheet in there at will until I got approval.

File: lib/static/css/common.css (hunk 3 above)
Comment (sholladay) [inline lib/static/css/common.css]: I personally find it easier to reason about sizing in additive terms. But I don't feel strongly about it.

If we are going to use this, we might want to use the [inherit approach](https://css-tricks.com/inheriting-box-sizing-probably-slightly-better-best-practice/).

File: lib/view/home-pages/fan-home.html (hunk 5 above)
Comment (sholladay) [inline lib/view/home-pages/fan-home.html]: My existing nav bar just doesn't work as the mockups call for. But it's the correct place to put the markup.

File: lib/static/css/common.css (hunk 3 above)
Comment (Bashkir15) [inline lib/static/css/common.css]: The inherit approach is a good solution to appeal to both. I'm comfortable with either, so we can base that on the rest of the team. For the most part I'm pretty flexible. I'll change this to the inherit approach for now and we can revist the topic on Saturday.

File: lib/view/home-pages/fan-home.html (hunk 5 above)
Comment (Bashkir15) [inline lib/view/home-pages/fan-home.html]: Yes, I definitely agree. I just didn't want to override that markup that existed and step on any toes haha. I will go and fix this now then.

File: lib/static/sass/config/typography.sass [hunk 6]
```diff
@@ -0,0 +1,5 @@
+@mixin open
```
Comment (sholladay) [inline lib/static/sass/config/typography.sass]: From what I can see, most of the Sass features you are using are native to CSS these days, such as variables and module imports. And I wouldn't use it for the syntax alone. But mixins are interesting.

I would like to hear the arguments for using it now, particularly the features you actually want to use that are _not_ in plain CSS.

Comment (Bashkir15) [inline lib/static/sass/config/typography.sass]: Preface: This a lot because I write like I talk: too much and with either way too much or way too little detail. There is a tl;dr. 

Well, given the scope of what I have done, I haven't really need to utilize many of the wonderful features of sass. The main thing is when it comes to nested properties and syntax, it is just more efficient to write and with tools like purecss and such, which I checked and there are loaders like this for rollup you can eliminate  99% of the bloat that might come with using it.

The main seller for me is the mixins and functions. It helps with DRY code so much especially when you are using keyframe animations throughout the application since I can just define a function that will take in number of steps, any animation properties and stuff like that, and prevent myself from writing out several lines of keyframes and also prevents me from having to import the entire animation file everytime, though again, that issue is negated by pure on compilation so the bloat isn't a big deal.

Also, media mixins. I didn't use them here because it wouldn't have been beneficial too considering there are so few pages built, but once we get a more comprehensive idea of how styles are going to universally translate across pages (these are the stress points for the design on all layouts, these is how our fonts and layout are going to develop around these stress points), it becomes much easier to develop media mixins for queries instead of  specifying breakpoints.

For example, I could define a media mixin (at this point basically just a media query that can take a variable and transform it), then use it across all our pages. Of course, there is nothing that is 100% universal in terms of design and where the design might be stressed. On one section the design my break at a screen width of 800px, but on another it falls apart at 870px. Instead of compromising the first slightly and switching the query up to 870 or defining way too many media queries , we can use a handy little mixin that takes an input that can respond dynamically to the browser viewport.

At very basics it is a handy way to define multiple media queries (which then can be combined and bloat reduced with a combine-media-queries plugin), and at its full power allows to have full control of the design dynamically on a pixel to pixel basis allowing us to interpolate in between stress points. 

All of this is "unnecessary" but is a huge quality of life improvement if done well. Also since the bloat can be almost entirely eliminated there isn't much cost.

Ultimately, it isn't going to matter because of the way you handle styles with React. The way I prefer to do it is similar to the way that Donovan described in our talk, but I tend to use multiple files. I will import a common sass/css/css-module into base component and then for each of my highest-level component that will handle conditional rendering I will import a sass/css/css-module file that delegates and namespaces the styles for all of the child components.

All of these features are of course support by rollup.

tl;dr 

1. I'm lazy and not being able to nest effectively along with enforced semi-colons and brackets are a huge pain

2. Mixins are incredibly handle and vital for doing things like animations where you have a high chance of repeating yourself and while this can be solved a little bit with css-modules, you can't import specific parts of a file (to my knowledge) and pass in arguments that could let you right 20 lines of code vs over a hundred.

3. functions, because I mean, we're programmers. These are kind of our life blood. The hyperbolic nature of that statement aside, working and maintaining a huge css codebase can be a nightmare. Modules make it easier, but are not yet a full-solution because I can write functions that will still reduce my code by a lot.

Ultimately, since a lot of browsers don't support import yet and we will have to use a bundle/build-tool for modules anyway, we might as well use sass (or postcss which in my opinion is better but I used Sass because more people are familiar with it) and have a main config and then import the styles as we need them inside the React components. 

Like I said though, I'm pretty flexible so I'm completely willing to work with anything the group decides is better, there in lies my decision to use some tech I like less but I feel people might be more familiar with, like gulp < 4 and sass, since I don't know much about the rest of the team in terms of what they have worked with and don't expect people to have to go and dive into a ton of documentation on my part haha,

Comment (Bashkir15) [inline lib/static/sass/config/typography.sass]: Commenting again instead of editing for visibility.

Here is a more concrete example. It is easy interpolate with length values using calc() and viewport height/width, mixins/function are handy for this, but not necessary.

A problem comes in with things like font-weight and font-size. Since calc() only works with lengths, font-weight brings in an issue because it is unitless.

I can attempt to solve this by defining a function using some principle from unit algebra to be able to do something like this

calc(2rem * 2rem) = 4

That could allow me to interpolate unitless values like font-weight or line-height if I wanted to make that unitless. I could even extend that out to non-linear equations to allow me multiply units by themselves. 

To further solve problem, since font-weight and style aren't exactly linear properties, you can make some fun of differential equations to create variable fonts given you have you have the proper encoding. Then you can situationally vary things like stroke-width etc, and even respond to things like zoom.

IFF (god why would you) you want to take things even further, you can start handling these calculations in the browser using Service workers and make these variable fonts respond to things like ambient light and viewing distance. This is much, much more complicated and I have only ever implemented a POC for the first. This is vastly outside of the scope of this project, but I was building on the established principle of why preprocessors are cool.

Comment (sholladay) [inline lib/static/sass/config/typography.sass]: > enforced semi-colons and brackets are a huge pain

My first language was Python, so I can relate. But go use CofeeScript for a while and I think you will change your mind. Or try writing English without punctuation. I'll take Java and its verbose `createFooFactory()` nonsense over one letter variable names any day of the week. When we create a formal style guide, it is going to include mandatory structural syntax. Our JavaScript linter already enforces this and I aim to do the same for stylesheets using [stylelint](https://github.com/stylelint/stylelint). Your IDE is there to assist you with such things.

> a lot of browsers don't support import yet

That's fine. We are Chrome only for the forseeable future. I will let you know months in advance if / when this changes. I would rather over-optimize for Chrome than deal with the weight and complexity of cross-browser issues before we need to.

> tech I like less but I feel people might be more familiar with

While that's a good trait, we are going to use the best tech, not the most popular. Hence Hapi > Express, RethinkDB > MongoDB, etc. We may even end up using [Riot](https://github.com/riot/riot) instead of React, though I'm not going to push that one too hard. But this is a theme that will continue.

Comment (Bashkir15) [inline lib/static/sass/config/typography.sass]: Haha, PEP8 bible and indent-errors galore. I never really get into CoffeScript, I used it a little but not too much. For me the strict-indentation and nesting enforcement that sass brings is my punctuation! Seriously though, preprocessor or no I'll manage. How valuable the mixins and functions are going to be remains to be seen and will stay the way until for a while until we decide what "polishing" features we are going to need. 

Stylelint makes me very happy. 


OH, that's completely fine I understand that, but Chrome doesn't natively support import yet either. 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import

It seems weird because its V8 just like Node and node supports a ton of newer features out of the box even without the --harmony flag. 


I'm completely fine with that. I actually prefer that. Though, and this is not me being divisive in any manner I'm just curious because you know more about Hapi than I do, what standard are we comparing by here? I will 100% agree with you on the Rethink v Mongo because you kind of get the best of both world with Rethink with the REQl query language, how easy it is to shard and create replica sets, external data-layers and all that yummy scaling stuff, but what benefits does Hapi posit? 

The reason I had never picked up happy before is because everyone had said it was slow in comparison to other, and it terms of handling sheer volume of requests it is slower, but obviously there is a lot more to a framework outside of arbitrary, not even that conclusive, benchmarks. So I'd love to learn more about the benefits and what makes it better, outside of the fact is more actively developed, which is definitely not something to ignore but I know that part. 

Well, hell, at that rate let's just go full on Inferno until React Fiber comes out. Mostly joking, but Inferno is neat.  Riot is also fun. I've used it on a couple of projects and wasn't disappointed. 


Also, I fixed the last requested change with the images by the way. Also, I'm sure you've noticed from the commits, I haven't added a route for this page. I'm not sure on the route names for some of these pages since there is a main home, a fan home, etc. I had just assumed /fans but I didn't want to mess up the scheme. 

Thanks for the quality questions/concerns by the way.
